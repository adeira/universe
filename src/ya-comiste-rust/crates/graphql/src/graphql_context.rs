#[derive(Clone)]
pub struct Context {
    pub pool: arangodb::ConnectionPool,
}

impl juniper::Context for Context {}
