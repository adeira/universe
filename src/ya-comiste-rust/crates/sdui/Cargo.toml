[package]
name = "sdui"
version = "0.0.0"
authors = ["Martin Zl√°mal <mrtnzlml@gmail.com>"]
edition = "2018"
publish = false

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
arangodb = { path = "../arangodb" }
arangors = "0.4.5"
graphql = { path = "../graphql" }
juniper = { git = "https://github.com/graphql-rust/juniper" }
serde = "1.0.117"
