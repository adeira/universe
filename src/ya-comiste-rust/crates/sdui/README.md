This package is responsible for managing Server Driven UI (SDUI).
