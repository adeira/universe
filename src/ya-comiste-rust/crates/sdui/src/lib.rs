pub mod entrypoint;
pub mod errors;
pub mod model;
pub mod sdui_section;
