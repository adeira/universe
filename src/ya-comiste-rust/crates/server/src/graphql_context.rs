#[derive(Clone, Copy, Debug)]
pub struct Context;
impl juniper::Context for Context {}
