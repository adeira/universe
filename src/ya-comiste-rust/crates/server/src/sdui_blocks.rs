use arangodb::connection;
use sdui::errors::ModelError;
use sdui::model::entrypoints::get_entrypoint;
use sdui::sdui_section::SDUISection;

pub async fn get_all_sections(entrypoint_key: String) -> Result<Vec<SDUISection>, ModelError> {
    let conn = connection().await;
    let db = conn.db("ya-comiste").await.unwrap();

    let entrypoint = get_entrypoint(&entrypoint_key).await?;
    let aql = arangors::AqlQuery::builder()
        .query(
            "
            FOR vertex IN 1..1 OUTBOUND @entrypoint_id entrypoint_sections
            FILTER vertex.type IN @supported_vertex_types
            SORT vertex.section_order ASC
            RETURN {
                id: vertex._id,
                title: vertex.title,
            }
            ",
        )
        .bind_var("entrypoint_id", entrypoint._id.to_string())
        .bind_var(
            "supported_vertex_types",
            vec!["SDUIScrollViewHorizontalComponent"], // TODO: filter by actual `supported` arg
        )
        .batch_size(1)
        .build();

    match db.aql_query::<SDUISection>(aql).await {
        Ok(r) => Ok(r),
        Err(e) => Err(ModelError::DatabaseError(e)),
    }
}
