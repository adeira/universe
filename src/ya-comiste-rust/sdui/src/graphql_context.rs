use crate::model::component_content::ContentDataloaderType;
use crate::user::User;

#[derive(Clone)]
pub struct Context {
    pub pool: arangodb::ConnectionPool,
    pub content_dataloader: ContentDataloaderType,
    pub user: User,
}

impl juniper::Context for Context {}
