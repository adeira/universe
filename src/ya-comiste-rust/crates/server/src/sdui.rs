use crate::graphql_context::Context;

/// Entrypoint represents a single entrypoint of the mobile application.
/// TODO: maybe `MobileEntrypoint` would be a better name so we are future proof?
#[derive(Clone, Debug, serde::Deserialize)]
pub struct SDUISection {
    pub id: juniper::ID,
    pub title: String,
}

// impl SDUISection {
//     fn not_exposed_to_graphql(&self) -> &str {
//         &self.title
//     }
// }

#[juniper::graphql_object(context = Context)]
impl SDUISection {
    #[graphql(description = "GraphQL description ðŸ¤¯")]
    fn id(&self) -> &str {
        &self.id
    }

    fn title(&self) -> &str {
        &self.title
    }
}
