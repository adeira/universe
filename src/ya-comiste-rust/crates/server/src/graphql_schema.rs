use crate::graphql_context::Context;
use crate::sdui::SDUISection;
use crate::sdui_blocks::get_all_blocks;
use juniper::{EmptyMutation, EmptySubscription, FieldError, RootNode};

#[derive(Clone, Copy, Debug)]
pub struct Query;

#[juniper::graphql_object(context = Context)]
impl Query {
    async fn mobile_entrypoint_sections(id: juniper::ID) -> Result<Vec<SDUISection>, FieldError> {
        let resp = get_all_blocks(id).await;
        Ok(resp)
    }
}

pub type Schema = RootNode<'static, Query, EmptyMutation<Context>, EmptySubscription<Context>>;

pub fn create_graphql_schema() -> Schema {
    Schema::new(
        Query,
        EmptyMutation::<Context>::new(),
        EmptySubscription::<Context>::new(),
    )
}

#[cfg(test)]
mod tests {
    #[test]
    fn schema_snapshot() {
        insta::assert_snapshot!(super::create_graphql_schema().as_schema_language());
    }
}
