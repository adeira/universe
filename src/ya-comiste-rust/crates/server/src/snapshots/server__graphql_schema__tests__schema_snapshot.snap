---
source: crates/server/src/graphql_schema.rs
expression: "super::create_graphql_schema().as_schema_language()"
---
type SDUIScrollViewHorizontalComponent {
  id: ID!
}

type SDUIDescriptionComponent {
  id: ID!
}

type SDUIJumbotronComponent {
  id: ID!
}

type SDUICardComponent {
  id: ID!
}

type Query {
  mobileEntrypointSections(key: String!): [SDUISection!]!
}

type SDUISection {
  id: ID!
  component(supported: [String!]!): SDUIComponent
}

union SDUIComponent = SDUICardComponent | SDUIDescriptionComponent | SDUIJumbotronComponent | SDUIScrollViewHorizontalComponent

schema {
  query: Query
}

